larq_compute_engine/core/bgemm_kernels_common.h

```diff
@@ -99,4 +99,54 @@ inline void MakeBinaryKernelParams(
   RUY_DCHECK_LT(params->last_col, params->dst_cols);
 }
 
+// A specialised template for the case when the LHS and RHS are uint32 bitpacked
+// but we're using a kernel designed for uint64 bitpacked inputs.
+template <int LhsCols, int RhsCols, typename AccumScalar>
+inline void MakeBinaryKernelParams(
+    const PackedMatrix<std::uint32_t>& lhs,
+    const PackedMatrix<std::uint32_t>& rhs,
+    const BinaryBasicSpec<AccumScalar, float>& spec, int start_row,
+    int start_col, int end_row, int end_col, Matrix<float>* dst,
+    BinaryKernelParams<LhsCols, RhsCols, std::uint64_t>* params) {
+  const int depth = lhs.layout.rows;
+  RUY_DCHECK_EQ(start_row % LhsCols, 0);
+  RUY_DCHECK_EQ(start_col % RhsCols, 0);
+  RUY_DCHECK_EQ(end_row % LhsCols, 0);
+  RUY_DCHECK_EQ(end_col % RhsCols, 0);
+
+  params->lhs_base_ptr = reinterpret_cast<std::uint64_t*>(
+      lhs.data + start_row * lhs.layout.stride);
+  params->rhs_base_ptr = reinterpret_cast<std::uint64_t*>(
+      rhs.data + start_col * rhs.layout.stride);
+  params->dst_base_ptr =
+      dst->data.get() + start_col * dst->layout.stride + start_row;
+
+  std::uint8_t flags = 0;
+  params->post_activation_multiplier =
+      spec.output_transform.post_activation_multiplier;
+  params->post_activation_bias = spec.output_transform.post_activation_bias;
+  if (params->post_activation_multiplier && params->post_activation_bias) {
+    flags |= RUY_ASM_FLAG_HAS_BIAS;
+  }
+  params->backtransform_add = spec.output_transform.backtransform_add;
+  params->flags = flags;
+  params->start_row = start_row;
+  params->start_col = start_col;
+  params->last_row = end_row - LhsCols;
+  params->last_col = end_col - RhsCols;
+  params->lhs_stride = sizeof(std::uint32_t) * lhs.layout.stride;
+  params->rhs_stride = sizeof(std::uint32_t) * rhs.layout.stride;
+  params->dst_stride = sizeof(float) * dst->layout.stride;
+  // We halve the depth to pretend that the input data is uint64.
+  RUY_DCHECK_EQ(depth % 2, 0);
```

Comment:
It's been some time since I've looked at this, but is `depth` measured in number of `TBitpacked` elements, or is it measured in number of "blocks" where a "block" is what is defined by `using LhsLayout = FixedKernelLayout<Order::kColMajor, 2, 4>;` ? I guess the first one.

What happens if the depth is odd? Does RUY's prepacking somehow make sure that doesn't happen because it knows about that block size thats expected by the kernel?

Reply:
It's the number of `TBitpacked` elements, and I believe that the Ruy pre-packing makes sure that that's not an issue by using padding (it knows the depth at compile-time as it's reported in the `FixedKernelLayout`, as you point out).